core/src/mast/mod.rs

```diff
@@ -0,0 +1,172 @@
+pub use call_node::CallNode;
+
+mod dyn_node;
+pub use dyn_node::DynNode;
+
+mod join_node;
+pub use join_node::JoinNode;
+
+mod split_node;
+pub use split_node::SplitNode;
+
+mod loop_node;
+pub use loop_node::LoopNode;
+
+pub trait MerkleTreeNode {
+    fn digest(&self) -> RpoDigest;
+}
+
+/// An opaque handle to a [`MastNode`] in some [`MastForest`]. It is the responsibility of the user
+/// to use a given [`MastNodeId`] with the corresponding [`MastForest`].
+///
+/// Note that since a [`MastForest`] enforces the invariant that equal [`MastNode`]s MUST have equal
+/// [`MastNodeId`]s, [`MastNodeId`] equality can be used to determine equality of the underlying
+/// [`MastNode`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct MastNodeId(usize);
+
+#[derive(Debug, Default)]
+pub struct MastForest {
+    /// All of the blocks local to the trees comprising the MAST forest
+    nodes: Vec<MastNode>,
+    node_id_by_hash: BTreeMap<RpoDigest, MastNodeId>,
+
+    /// The "entrypoint", when set, is the root of the entire forest, i.e.
+    /// a path exists from this node to all other roots in the forest. This
+    /// corresponds to the executable entry point. When not set, the forest
+    /// may or may not have such a root in `roots`, but is not required.
+    /// Whether or not the entrypoint is set distinguishes a MAST which is
+    /// executable, versus a MAST which represents a library.
+    ///
+    /// NOTE: The entrypoint is also present in `roots` if set
+    entrypoint: Option<MastNodeId>,
+    kernel: Kernel,
+}
+
+impl MastForest {
+    /// Creates a new empty [`MastForest`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Adds a node to the forest, and returns the [`MastNodeId`] associated with it.
+    ///
+    /// If a [`MastNode`] which is equal to the current node was previously added, the previously
+    /// returned [`MastNodeId`] will be returned. This enforces this invariant that equal
+    /// [`MastNode`]s have equal [`MastNodeId`]s.
+    pub fn add_node(&mut self, node: MastNode) -> MastNodeId {
+        let node_digest = node.digest();
+
+        if let Some(node_id) = self.node_id_by_hash.get(&node_digest) {
+            // node already exists in the forest; return previously assigned id
+            *node_id
+        } else {
+            let new_node_id = MastNodeId(self.nodes.len());
+
+            self.node_id_by_hash.insert(node.digest(), new_node_id);
+            self.nodes.push(node);
+
+            new_node_id
+        }
+    }
+
+    pub fn kernel(&self) -> &Kernel {
+        &self.kernel
+    }
+
+    pub fn entrypoint(&self) -> Option<MastNodeId> {
+        self.entrypoint
+    }
+
+    /// A convenience method that provides the hash of the entrypoint, if any.
+    pub fn entrypoint_digest(&self) -> Option<RpoDigest> {
+        self.entrypoint.map(|entrypoint| self.get_node_by_id(entrypoint).digest())
+    }
+
+    pub fn get_node_by_id(&self, node_id: MastNodeId) -> &MastNode {
```

Comment:
**Suggestion:** Annotate this with `#[inline(always)]`

Reply:
**Suggestion:** Implement `Index` and optionally `IndexMut` for `MastForest` with `MastNodeId`, since that will be a super common operation.

Reply:
Implemented `Index`, but IMO we should either have `Index`, or `get_node_by_id()`, but not both, as it leads to both being used at different places in the codebase. Remembering that they're equivalent when parsing the code is unneeded mental overhead.

I don't have a strong preference for either; what do you think?

Reply:
If you have both, usually the convention is to have the `get` method be infallible (i.e. return an `Option`/`Result`), since the semantics of `Index` don't really permit that.